spread/openstack.go

```diff
@@ -0,0 +1,631 @@
+
+func (s *openstackServer) Address() string {
+	return s.address
+}
+
+func (s *openstackServer) System() *System {
+	return s.system
+}
+
+func (s *openstackServer) ReuseData() interface{} {
+	return &s.d
+}
+
+const (
+	openstackStaging      = "STAGING"
+	openstackProvisioning = "PROVISIONING"
+	openstackRunning      = "RUNNING"
+	openstackStopping     = "STOPPING"
+	openstackStopped      = "STOPPED"
+	openstackSuspending   = "SUSPENDING"
+	openstackTerminating  = "TERMINATED"
+
+	openstackPending = "PENDING"
+	openstackDone    = "DONE"
+)
+
+func (p *openstackProvider) Backend() *Backend {
+	return p.backend
+}
+
+func (p *openstackProvider) Reuse(ctx context.Context, rsystem *ReuseSystem, system *System) (Server, error) {
+	s := &openstackServer{
+		p:       p,
+		address: rsystem.Address,
+		system:  system,
+	}
+	err := rsystem.UnmarshalData(&s.d)
+	if err != nil {
+		return nil, fmt.Errorf("cannot unmarshal openstack reuse data: %v", err)
+	}
+	return s, nil
+}
+
+func (p *openstackProvider) Allocate(ctx context.Context, system *System) (Server, error) {
+	if err := p.checkKey(); err != nil {
+		return nil, err
+	}
+
+	s, err := p.createMachine(ctx, system)
+	if err != nil {
+		return nil, err
+	}
+
+	printf("Allocated %s.", s)
+	return s, nil
+}
+
+func (s *openstackServer) Discard(ctx context.Context) error {
+	return s.p.removeMachine(ctx, s)
+}
+
+const openstackCloudInitScript = `
+#cloud-config
+runcmd:
+  - echo root:%s | chpasswd
+  - sed -i 's/^\s*#\?\s*\(PermitRootLogin\|PasswordAuthentication\)\>.*/\1 yes/' /etc/ssh/sshd_config
+  - test -d /etc/ssh/sshd_config.d && echo 'PermitRootLogin=yes' > /etc/ssh/sshd_config.d/00-spread.conf
+  - test -d /etc/ssh/sshd_config.d && echo 'PasswordAuthentication=yes' >> /etc/ssh/sshd_config.d/00-spread.conf
+  - pkill -o -HUP sshd || true
+`
+const openstackReadyMarker = "MACHINE-IS-READY"
+const openstackNameLayout = "Jan021504.000000"
+const openstackDefaultFlavor = "m1.medium"
+
+type openstackImage struct {
+	Project string
+	Name    string
+	Family  string
+	Terms   []string
+}
+
+type openstackImagesCache struct {
+	mu     sync.Mutex
+	ready  bool
+	images []openstackImage
+	err    error
+}
+
+func openstackName() string {
+	return strings.ToLower(strings.Replace(time.Now().UTC().Format(openstackNameLayout), ".", "-", 1))
+}
+
+func findErrorTitle(node *html.Node) string {
+	if node.Type == html.ElementNode && node.Data == "title" {
+		return node.FirstChild.Data
+	}
+
+	for child := node.FirstChild; child != nil; child = child.NextSibling {
+		title := findErrorTitle(child)
+		if title != "" {
+			return title
+		}
+	}
+
+	return ""
+}
+
+// error messages returned by openstack api are html
+// which contain title and details related to the error
+func errorTitle(msg string) string {
+	node, err := html.Parse(strings.NewReader(msg))
+	if err != nil {
+		return ""
+	}
+	return findErrorTitle(node)
+}
+
+func (p *openstackProvider) findFlavor(flavorName string) (nova.Entity, error) {
```

Comment:
(nitpick) it seems to me that using a pointer to nova.Entity here would be slightly better as you can return `nil` instead of an "empty" nova.Entity. Plus when found just returning the value instead of break and indirect return is probably slightly simpler. 

I.e.:
```go
diff --git a/spread/openstack.go b/spread/openstack.go
index 52c6aaf..417e67a 100644
--- a/spread/openstack.go
+++ b/spread/openstack.go
@@ -206,26 +206,19 @@ func errorTitle(msg string) string {
        return findErrorTitle(node)
 }
 
-func (p *openstackProvider) findFlavor(flavorName string) (nova.Entity, error) {
-       var flavor nova.Entity
-
+func (p *openstackProvider) findFlavor(flavorName string) (*nova.Entity, error) {
        flavors, err := p.computeClient.ListFlavors()
        if err != nil {
-               return flavor, fmt.Errorf("failed to retrieve flavors list: %s", errorTitle(err.Error()))
+               return nil, fmt.Errorf("failed to retrieve flavors list: %s", errorTitle(err.Error()))
        }
 
        for _, f := range flavors {
                if f.Name == flavorName {
-                       flavor = f
-                       break
+                       return &f, nil
                }
        }
 
-       if flavor.Id == "" {
-               return flavor, fmt.Errorf("specified flavor not found: %s", flavorName)
-       }
-
-       return flavor, nil
+       return nil, fmt.Errorf("specified flavor not found: %s", flavorName)
 }
 
 func (p *openstackProvider) findNetwork(name string) (neutron.NetworkV2, error) {
```